File: litfass-server/src/main/scala/lit/fass/server/script/ScriptActor.scala
package lit.fass.server.script

import akka.actor.typed.scaladsl.{AbstractBehavior, ActorContext, Behaviors}
import akka.actor.typed.{ActorRef, Behavior}
import lit.fass.server.ScriptLanguage.ScriptLanguage
import lit.fass.server.model.DataList
import lit.fass.server.script.ScriptActor.{ExecuteScriptMessage, ExecutedScriptMessage, ScriptMessage}


object ScriptActor {

  //@formatter:off
  sealed trait ScriptMessage
  final case class ExecuteScriptMessage(language: ScriptLanguage, script: String, data: DataList, replyTo: ActorRef[ExecutedScriptMessage]) extends ScriptMessage
  final case class ExecutedScriptMessage(data: DataList)
  //@formatter:on

  def apply(): ScriptActor =
    Behaviors.setup(context => new ScriptActor(context = context))

  def apply(scriptEngines: List[ScriptEngine]): ScriptActor =
    Behaviors.setup(context => new ScriptActor(scriptEngines, context))
}

/**
 * @author Michael Mair
 */
class ScriptActor(private val scriptEngines: List[ScriptEngine] = List(GroovyScriptEngine()),
                  context: ActorContext[ScriptMessage]) extends AbstractBehavior[ScriptMessage](context) {


  override def onMessage(msg: ScriptMessage): Behavior[ScriptMessage] = {
    msg match {
      case ExecuteScriptMessage(lang, script, data, replyTo) =>
        val engine = scriptEngines.find { engine => engine.isApplicable(lang) }.orNull
        if (engine == null) throw new ScriptException(s"No script engine available for language $language")
        val result = engine.invoke(script, data)
        replyTo ! ExecutedScriptMessage(result)
        this
    }
  }
}

File: litfass-server/src/main/kotlin/lit/fass/server/actor/ScriptActor.kt
package lit.fass.server.actor

import akka.actor.typed.ActorRef
import akka.actor.typed.Behavior
import akka.actor.typed.javadsl.AbstractBehavior
import akka.actor.typed.javadsl.ActorContext
import akka.actor.typed.javadsl.Behaviors
import akka.actor.typed.javadsl.Behaviors.same
import akka.actor.typed.javadsl.Receive
import lit.fass.server.actor.ScriptActor.Message
import lit.fass.server.script.ScriptEngine
import lit.fass.server.script.ScriptException
import lit.fass.server.script.ScriptLanguage
import java.util.*


/**
 * @author Michael Mair
 */
class ScriptActor private constructor(
    private val scriptEngines: List<ScriptEngine>,
    context: ActorContext<Message>?
) : AbstractBehavior<Message>(context) {

    companion object {
        @JvmStatic
        fun create(scriptEngines: List<ScriptEngine>): Behavior<Message> {
            return Behaviors.setup { context -> ScriptActor(scriptEngines, context) }
        }
    }

    interface Message : SerializationMarker
    data class TestScript(val language: ScriptLanguage, val payload: Map<*, *>, val replyTo: ActorRef<ScriptResult>) : Message
    data class ScriptResult(val result: Collection<Map<String, Any?>>) : Message

    override fun createReceive(): Receive<Message> = newReceiveBuilder()
        .onMessage(TestScript::class.java) {
            val scriptEngine = scriptEngines.find { engine -> engine.isApplicable(it.language) }
                ?: throw ScriptException("No script engine available for language ${it.language}")

            val script = it.payload["script"] as String

            @Suppress("UNCHECKED_CAST")
            val data = it.payload["data"] as Map<String, Any?>

            val result = scriptEngine.invoke(script, Collections.singletonList(data))
            it.replyTo.tell(ScriptResult(result))
            same()
        }
        .build()

}